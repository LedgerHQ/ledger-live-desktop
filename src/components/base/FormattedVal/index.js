// @flow

import React from 'react'
import styled from 'styled-components'
import { connect } from 'react-redux'

import isUndefined from 'lodash/isUndefined'

import type { Unit } from '@ledgerhq/live-common/lib/types'
import type { State } from 'reducers'

import {
  formatCurrencyUnit,
  findCurrencyByTicker,
} from '@ledgerhq/live-common/lib/helpers/currencies'

import { marketIndicatorSelector, localeSelector } from 'reducers/settings'

import { getMarketColor } from 'styles/helpers'

import Box from 'components/base/Box'
import FlipTicker from 'components/base/FlipTicker'

import IconBottom from 'icons/Bottom'
import IconTop from 'icons/Top'

const T = styled(Box).attrs({
  ff: 'Rubik',
  horizontal: true,
  color: p => (p.withIcon ? p.theme.colors.dark : p.color),
})`
  line-height: 1.2;
  white-space: pre;
`

const I = ({ color, children }: { color?: string, children: any }) => (
  <Box color={color}>{children}</Box>
)

I.defaultProps = {
  color: undefined,
}

type OwnProps = {
  alwaysShowSign?: boolean,
  animateTicker?: boolean,
  color?: string,
  disableRounding?: boolean,
  fiat?: string,
  isPercent?: boolean,
  showCode?: boolean,
  unit?: Unit,
  val: number,
  withIcon?: boolean,
}

const mapStateToProps = (state: State, _props: OwnProps) => ({
  marketIndicator: marketIndicatorSelector(state),
  locale: localeSelector(state),
})

type Props = OwnProps & {
  marketIndicator: string,
  locale: string,
}

let _logged

function FormattedVal(props: Props) {
  const {
    animateTicker,
    disableRounding,
    fiat,
    isPercent,
    alwaysShowSign,
    showCode,
    withIcon,
    locale,
    marketIndicator,
    color,
    ...p
  } = props
  let { val, unit } = props

  if (isUndefined(val)) {
    throw new Error('FormattedVal require a `val` prop. Received `undefined`')
  }

  const isNegative = val < 0

  let text = ''

  if (isPercent) {
    text = `${alwaysShowSign ? (isNegative ? '- ' : '+ ') : ''}${isNegative ? val * -1 : val} %`
  } else {
    if (fiat) {
      if (!_logged) {
        _logged = true
        console.warn('FormattedVal: passing fiat prop is deprecated')
      }
      const cur = findCurrencyByTicker(fiat)
      if (cur) {
        ;[unit] = cur.units
      }
    }
    if (!unit) {
      return ''
    }

    if (withIcon && isNegative) {
      val *= -1
    }

    text = formatCurrencyUnit(unit, val, {
      alwaysShowSign,
      disableRounding,
      showCode,
      locale,
    })
  }

  if (animateTicker) {
    text = <FlipTicker value={text} />
  }

  const marketColor = getMarketColor({
    marketIndicator,
    isNegative,
  })

  return (
    <T color={color || marketColor} withIcon={withIcon} {...p}>
      {withIcon ? (
        <Box horizontal alignItems="center" flow={1}>
          <Box>
            <I color={marketColor}>
              {isNegative ? <IconBottom size={16} /> : <IconTop size={16} />}
            </I>
          </Box>
          <Box horizontal alignItems="center">
            {text}
          </Box>
        </Box>
      ) : (
        text
      )}
    </T>
  )
}

export default connect(mapStateToProps)(FormattedVal)
