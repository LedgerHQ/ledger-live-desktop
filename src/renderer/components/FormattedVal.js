// @flow

import type { BigNumber } from "bignumber.js";
import invariant from "invariant";
import React from "react";
import styled from "styled-components";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import type { Unit } from "@ledgerhq/live-common/lib/types";
import { formatCurrencyUnit } from "@ledgerhq/live-common/lib/currencies";
import {
  marketIndicatorSelector,
  localeSelector,
  discreetModeSelector,
} from "~/renderer/reducers/settings";
import { getMarketColor } from "~/renderer/styles/helpers";
import Box from "~/renderer/components/Box";
import FlipTicker from "~/renderer/components/FlipTicker";
import IconBottom from "~/renderer/icons/ArrowDownRight";
import IconTop from "~/renderer/icons/ArrowUpRight";
import Ellipsis from "~/renderer/components/Ellipsis";

import type { ThemedComponent } from "~/renderer/styles/StyleProvider";

const T: ThemedComponent<{ color?: string, inline?: boolean, ff?: string }> = styled(Box).attrs(
  p => ({
    ff: p.ff || "Inter|Medium",
    horizontal: true,
    color: p.color,
  }),
)`
  white-space: pre;
  text-overflow: ellipsis;
  display: ${p => (p.inline ? "inline-block" : "block")};
  flex-shrink: 1;
  width: ${p => (p.inline ? "" : "100%")};
  overflow: hidden;
`;

const I = ({ color, children }: { color?: string, children: any }) => (
  <Box color={color}>{children}</Box>
);

I.defaultProps = {
  color: undefined,
};

type OwnProps = {
  unit?: Unit,
  val: BigNumber,
  alwaysShowSign?: boolean,
  showCode?: boolean,
  withIcon?: boolean,
  color?: string,
  animateTicker?: boolean,
  disableRounding?: boolean,
  isPercent?: boolean,
  subMagnitude?: number,
  prefix?: string,
  ellipsis?: boolean,
  suffix?: string,
  showAllDigits?: boolean,
};

const mapStateToProps = createStructuredSelector({
  marketIndicator: marketIndicatorSelector,
  discreet: discreetModeSelector,
  locale: localeSelector,
});

type Props = OwnProps & {
  marketIndicator: string,
  discreet: boolean,
  locale: string,
};

function FormattedVal(props: Props) {
  const {
    animateTicker,
    disableRounding,
    unit,
    isPercent,
    alwaysShowSign,
    showCode,
    withIcon,
    locale,
    marketIndicator,
    color,
    ellipsis,
    subMagnitude,
    prefix,
    suffix,
    discreet,
    showAllDigits,
    ...p
  } = props;
  let { val } = props;

  invariant(val, "FormattedVal require a `val` prop. Received `undefined`");

  const isNegative = val.isNegative() && !val.isZero();

  let text = "";

  if (isPercent) {
    // FIXME move out the % feature of this component... totally unrelated to currency & annoying for flow type.
    text = `${alwaysShowSign ? (isNegative ? "- " : "+ ") : ""}${(isNegative
      ? val.negated()
      : val
    ).toString()} %`;
  } else {
    invariant(unit, "FormattedVal require a `unit` prop. Received `undefined`");

    if (withIcon && isNegative) {
      val = val.negated();
    }

    text = formatCurrencyUnit(unit, val, {
      alwaysShowSign,
      disableRounding,
      showCode,
      locale,
      subMagnitude,
      discreet,
      showAllDigits,
    });
  }

  if (prefix) text = prefix + text;
  if (suffix) text += suffix;

  if (animateTicker) {
    text = <FlipTicker value={text} />;
  } else if (ellipsis) {
    text = <Ellipsis>{text}</Ellipsis>;
  }

  const marketColor = getMarketColor({
    marketIndicator,
    isNegative,
  });

  return (
    <T {...p} color={color || marketColor} withIcon={withIcon}>
      {withIcon ? (
        <Box horizontal alignItems="center">
          <Box mr={1}>
            <I color={marketColor}>
              {isNegative ? <IconBottom size={24} /> : <IconTop size={24} />}
            </I>
          </Box>
          <Box horizontal alignItems="center">
            {text}
          </Box>
        </Box>
      ) : (
        text
      )}
    </T>
  );
}

FormattedVal.defaultProps = {
  subMagnitude: 0,
};

const m: React$ComponentType<OwnProps> = connect(mapStateToProps)(FormattedVal);

export default m;
