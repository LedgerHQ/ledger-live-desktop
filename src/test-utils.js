import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'

import theme from 'styles/theme'

export function render(component) {
  return renderer.create(<ThemeProvider theme={theme}>{component}</ThemeProvider>).toJSON()
}
