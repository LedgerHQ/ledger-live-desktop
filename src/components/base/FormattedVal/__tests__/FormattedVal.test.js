import React from 'react'
import { getDefaultUnitByCoinType } from '@ledgerhq/currencies'

import { render } from 'test-utils'
import FormattedVal from '..'

const bitcoinUnit = getDefaultUnitByCoinType(0)

describe('components', () => {
  describe('FormattedVal', () => {
    it('renders a formatted val', () => {
      const component = <FormattedVal unit={bitcoinUnit} val={400000000} />
      const tree = render(component)
      expect(tree).toMatchSnapshot()
    })

    it('shows sign', () => {
      const component = <FormattedVal alwaysShowSign unit={bitcoinUnit} val={400000000} />
      const tree = render(component)
      expect(tree).toMatchSnapshot()

      const component2 = <FormattedVal alwaysShowSign unit={bitcoinUnit} val={-400000000} />
      const tree2 = render(component2)
      expect(tree2).toMatchSnapshot()
    })

    it('shows code', () => {
      const component = <FormattedVal showCode unit={bitcoinUnit} val={400000000} />
      const tree = render(component)
      expect(tree).toMatchSnapshot()
    })

    it('renders a percent', () => {
      const component = <FormattedVal isPercent val={30} />
      const tree = render(component)
      expect(tree).toMatchSnapshot()
    })

    it('renders a fiat', () => {
      const component = <FormattedVal fiat="USD" val={20} />
      const tree = render(component)
      expect(tree).toMatchSnapshot()
    })
  })
})
