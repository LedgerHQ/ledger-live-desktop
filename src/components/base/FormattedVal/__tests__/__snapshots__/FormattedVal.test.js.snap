// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components FormattedVal renders a fiat 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 eoitYS"
  color="#66be54"
>
  20.00
</div>
`;

exports[`components FormattedVal renders a formatted val 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 eoitYS"
  color="#66be54"
>
  4
</div>
`;

exports[`components FormattedVal renders a percent 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 eoitYS"
  color="#66be54"
>
  30 %
</div>
`;

exports[`components FormattedVal shows code 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 eoitYS"
  color="#66be54"
>
  BTC 4
</div>
`;

exports[`components FormattedVal shows sign 1`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 eoitYS"
  color="#66be54"
>
  + 4
</div>
`;

exports[`components FormattedVal shows sign 2`] = `
<div
  className="k45ou1-0 iqaJGf e345n3-0 lkKRAD"
  color="#ea2e49"
>
  - 4
</div>
`;
